pylops/signalprocessing/dtcwt.py

```diff
@@ -0,0 +1,145 @@
+
+
+class DTCWT(LinearOperator):
+    r"""
+    Perform Dual-Tree Complex Wavelet Transform on a given array.
+
+    This operator wraps around :py:func:`dtcwt` package.
+
+    Parameters
+    ----------
+    dims: :obj:`int` or :obj:`tuple`
+        Number of samples for each dimension.
+    transform: :obj:`int`, optional
+        Type of transform 1D, 2D or 3D. Default is 1.
+    birot: :obj:`str`, optional
+        Level 1 wavelets to use. See :py:func:`dtcwt.coeffs.birot()`. Default is `"near_sym_a"`.
+    qshift: :obj:`str`, optional
+        Level >= 2 wavelets to use. See :py:func:`dtcwt.coeffs.qshift()`. Default is `"qshift_a"`
+    nlevels: :obj:`int`, optional
+        Number of levels of wavelet decomposition. Default is 3.
+    include_scale: :obj:`bool`, optional
+        Include scales in pyramid. See :py:func:`dtcwt.Pyramid`. Default is False.
+    dtype : :obj:`DTypeLike`, optional
+        Type of elements in input array.
+    name : :obj:`str`, optional
+        Name of operator (to be used by :func:`pylops.utils.describe.describe`)
+
+    Raises
+    ------
+    NotImplementedError
+        If ``transform`` is 2 or 3.
+    ValueError
+        If ``transform`` is anything other than 1, 2 or 3.
+
+    Notes
+    -----
+    The :py:func:`dtcwt` library uses a Pyramid object to represent the transform domain signal.
+    It has
+        `lowpass` (coarsest scale lowpass signal)
+        `highpasses` (complex subband coefficients for corresponding scales)
+        `scales` (lowpass signal for corresponding scales finest to coarsest)
+
+    To make the dtcwt forward() and inverse() functions compatible with pylops, the Pyramid object is
+    flattened out and all coeffs(highpasses and low pass signal) are appened into one array using the
+    `_coeff_to_array` method.
+    For inverse, the flattened array is used to reconstruct the Pyramid object using the `_array_to_coeff`
+    method and then inverse is performed.
+
+    """
+
+    def __init__(
+        self,
+        dims: Union[int, InputDimsLike],
+        transform: int = 1,
+        biort: str = "near_sym_a",
+        qshift: str = "qshift_a",
+        nlevels: int = 3,
+        include_scale: bool = False,
+        dtype: DTypeLike = "float64",
+        name: str = "C",
+    ) -> None:
+        self.dims = _value_or_sized_to_tuple(dims)
+        self.transform = transform
+        self.ndim = len(self.dims)
+        self.nlevels = nlevels
+        self.include_scale = include_scale
+
+        if self.transform == 1:
+            self._transform = dtcwt.Transform1d(biort=biort, qshift=qshift)
+        elif self.transform == 2:
+            raise NotImplementedError("DTCWT is not implmented for 2D")
+        elif self.transform == 3:
+            raise NotImplementedError("DTCWT is not implmented for 3D")
+        else:
+            raise ValueError("DTCWT only supports 1D, 2D and 3D")
+
+        pyr = self._transform.forward(
```

Comment:
Ideally I would hide this away into a method (eg., `_interpret_coeffs`): from line 88 to 102.

Even better, I would see if you can avoid having a dry-run of dtcwt.Transform1d... if they dont provide any routine for that we can stick with this strategy for now.

Reply:
separated it out in a function [f887031](https://github.com/PyLops/pylops/pull/495/commits/f887031ff0aed0edac70bed9bf1388e2dbfb4b70)